fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/UnionToValuesRule.java

```diff
@@ -0,0 +1,167 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+// Transform all constant UNION ALLs into a single Values operator
+//
+// Pattern:
+//
+//      Union
+//      /   \
+//     X     Y
+//
+// Transform:
+//
+//           Union                              Union
+//         /   |   \          ===>             /     \
+//   Values1 Child2 Values3               Child2     Values(1, 3)
+//
+//
+//            UnionAll
+//         /     |     \          ===>      Values(1, 2, 3)
+//   Values1  Value2   Values3
+//
+//
+// Requirements:
+// 1. The operators for x and y must be of the LOGICAL_VALUES type, and must not contain limit and Predicate
+
+public class UnionToValuesRule extends TransformationRule {
+
+    private UnionToValuesRule() {
+        super(RuleType.TF_MERGE_CONSTANT_UNION, Pattern.create(OperatorType.LOGICAL_UNION)
+                .addChildren(Pattern.create(OperatorType.PATTERN_MULTI_LEAF)));
+    }
+
+    public static UnionToValuesRule getInstance() {
+        return INSTANCE;
+    }
+
+    private static final UnionToValuesRule INSTANCE = new UnionToValuesRule();
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        return input.getInputs().stream()
+                .filter(UnionToValuesRule::isMergeable)
+                .count() > 1;
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        LogicalUnionOperator unionOp = (LogicalUnionOperator) input.getOp();
+
+        List<List<ScalarOperator>> newRows = new ArrayList<>();
+        List<OptExpression> otherChildren = new ArrayList<>();
+        List<List<ColumnRefOperator>> newChildOutputs = Lists.newArrayList();
+        int firstLogicalValuesIndex = -1;
+
+        for (int i = 0; i < input.getInputs().size(); i++) {
+            OptExpression child = input.getInputs().get(i);
+            if (isMergeable(child)) {
+                if (firstLogicalValuesIndex == -1) {
+                    firstLogicalValuesIndex = i;
+                }
+
+                LogicalValuesOperator valuesOp = (LogicalValuesOperator) child.getOp();
+                List<List<ScalarOperator>> rows = valuesOp.getRows();
+                if (isConstantUnion(valuesOp)) {
+                    List<ScalarOperator> scalarOperators = unionOp.getChildOutputColumns().get(i).stream()
+                            .map(valuesOp.getProjection().getColumnRefMap()::get).collect(Collectors.toList());
+                    newRows.add(scalarOperators);
+                } else if (isConstantValues(valuesOp)) {
+                    newRows.addAll(rows);
+                }
+
+            } else {
+                newChildOutputs.add(unionOp.getChildOutputColumns().get(i));
+                otherChildren.add(child);
+            }
+        }
+
+        if (otherChildren.isEmpty() && !newRows.isEmpty() && unionOp.isUnionAll()) {
```

Comment:
If it's a not union all union which means we need deduplicate the rows, we need a agg operator here.

Reply:
syntax `union` will generate `agg-union`, so I think `LogicalUnionOperator` must be `union all`, may be we can remove  `isUnionAll` from `LogicalUnionOperator`

Reply:
You're right.